clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -1305,44 +1314,62 @@ private Sensor sensor(String name, Sensor... parents) {
-                    nodeRequest.add(createMeter(metrics, perConnectionMetricGrpName, tags, new WindowedCount(), "request", "requests sent"));
-                    MetricName metricName = metrics.metricName("request-size-avg", perConnectionMetricGrpName, "The average size of requests sent.", tags);
-                    nodeRequest.add(metricName, new Avg());
-                    metricName = metrics.metricName("request-size-max", perConnectionMetricGrpName, "The maximum size of any request sent.", tags);
-                    nodeRequest.add(metricName, new Max());
-
-                    String bytesSentName = "node-" + connectionId + ".bytes-sent";
-                    Sensor bytesSent = sensor(bytesSentName);
-                    bytesSent.add(createMeter(metrics, perConnectionMetricGrpName, tags, "outgoing-byte", "outgoing bytes"));
-
-                    String nodeResponseName = "node-" + connectionId + ".responses-received";
-                    Sensor nodeResponse = sensor(nodeResponseName);
-                    nodeResponse.add(createMeter(metrics, perConnectionMetricGrpName, tags, new WindowedCount(), "response", "responses received"));
-
-                    String bytesReceivedName = "node-" + connectionId + ".bytes-received";
-                    Sensor bytesReceive = sensor(bytesReceivedName);
-                    bytesReceive.add(createMeter(metrics, perConnectionMetricGrpName, tags, "incoming-byte", "incoming bytes"));
-
-                    String nodeTimeName = "node-" + connectionId + ".latency";
-                    Sensor nodeRequestTime = sensor(nodeTimeName);
-                    metricName = metrics.metricName("request-latency-avg", perConnectionMetricGrpName, tags);
-                    nodeRequestTime.add(metricName, new Avg());
-                    metricName = metrics.metricName("request-latency-max", perConnectionMetricGrpName, tags);
-                    nodeRequestTime.add(metricName, new Max());
-                }
+            if (!connectionId.isEmpty() && connectionMetrics != null) {
+                connectionMetrics.computeIfAbsent(connectionId, (key) -> {
+                    // key: connection id
+                    // value: set of sensors (currently null)
+                    return perConnectionSensors(key);
+                });
             }
         }
 
+        public void maybeUnregisterConnectionMetrics(String connectionId) {
+            if (!connectionId.isEmpty() && connectionMetrics != null) {
+                connectionMetrics.computeIfPresent(connectionId, (key, value) -> {
+                    // key: connection id
+                    // value: set of sensors
+                    for (Sensor sensor : value) {
+                        metrics.removeSensor(sensor.name());
+                    }
+                    return null;
+                });
+            }
+        }
+
+        private Set<Sensor> perConnectionSensors(String connectionId) {
```

Comment:
Would it make sense to create a `ConnectionMetrics` class to hold all the connection metrics? That would give us an opportunity to improve all the `record*` methods as well. They could get the sensors based on the `connectionId`.